File: src/main/scala/pl/msulima/vistula/template/transpiler/Expression.scala
package pl.msulima.vistula.template.transpiler

import pl.msulima.vistula.parser.Ast
import pl.msulima.vistula.transpiler.{CodeTemplate, RxMap}

import scala.io.Source

object Expression {

  private val MagicInlineHtmlPrefix = "# html\n"
  private val MagicClasspathHtmlRegex = "^# html:(.+?)".r

  def apply: PartialFunction[Ast.expr, CodeTemplate] = {
    case Ast.expr.Str(MagicClasspathHtmlRegex(sourceFile)) =>
      val stream = getClass.getResourceAsStream(sourceFile)
      require(stream != null, s"$sourceFile not found")
      val source = Source.fromInputStream(stream)
      val lines = source.getLines().toList
      source.close()

      CodeTemplate(Template(lines.mkString("\n")), RxMap)
    case Ast.expr.Str(x) if x.startsWith(MagicInlineHtmlPrefix) =>
      CodeTemplate(Template(x.stripPrefix(MagicInlineHtmlPrefix)), RxMap)
  }
}

File: src/test/resources/pl/msulima/vistula/transpiler/expression.js
W;
const P = Y.rxMap($arg => ($arg - 2 + $arg));
const Q = vistula.constantObservable(2 + 3);
const X = vistula.constantObservable(2 + 3 + 4);
const Y = a(Y, vistula.constantObservable(3));
const Z = vistula.zip([
    Y,
    a(Z.rxMap($arg => ($arg + 1)), vistula.constantObservable(3))
]).rxMap($args => ($args[0] + 3 - $args[1]));
